### ISSUE TITLE
Close Custom Tabs on Android on Result

### ISSUE THREAD

**Issue by DanielEberl** [teranetsrl/oauth2_client#26]
On Android, the Chrome / Browser Window stays open after the successfull redirect back to the app.

In some cases, when logging in again, the same instance gets opened, resulting in an unusable screen.

Is there an alternative to using custom tabs? Like webview i.e.?

**Comment by okrad** [comment]
Hi @DanielEberl, the OAuth rfc advises against using webviews, mainly for security reasons (such as avoiding the app to modify or inspect the browser window).

I can't reproduce your issue... Can you provide more info, such as:
* Android/Chrome version
* the service you are trying to authorize to
* a code snippet

**Comment by htinlinnzaw** [comment]
Hi, I am facing same issue. After Authorization success, the Chrome browser does not close automatically.

**Comment by htinlinnzaw** [comment]
My Chrome Version: com.android.chrome 84.0.4147.125

**Comment by okrad** [comment]
Hi @htinlinnzaw, 
I think the problem is caused by the issue mentioned above in flutter_web_auth, the library oauth2_client relies upon to handle the browser instance. Hope it will get fixed soon!

**Comment by ThexXTURBOXx** [comment]
I will just copy-paste my answer here, as users may find the answer faster like this:

I don't know, if the developer of flutter_web_auth is still actively developing his project. For the time being, you can just use my patch using a override, like this (add it to your pubspec.yaml):
```
dependency_overrides:
  flutter_web_auth:
    git: 'https://github.com/ThexXTURBOXx/flutter_web_auth.git'
```
Please be aware, that compile issues complaining about ``duplicate files`` may arise. I don't know exactly, why Dart/Flutter is too dumb to see, that my library should override all files of the other one (if someone has a fix for this problem, you are invited to tell me and all the others :) ).
My workaround for this is to just delete these folders:
``$FLUTTER_DIR$\.pub-cache\git\flutter_web_auth-<SOMEHASH>``
``$FLUTTER_DIR$\.pub-cache\hosted\pub.dartlang.org\flutter_web_auth-<SOMEHASH>``
and then ``pub get`` again.

**Comment by htinlinnzaw** [comment]
> Hi @htinlinnzaw,
> I think the problem is caused by the issue mentioned above in flutter_web_auth, the library oauth2_client relies upon to handle the browser instance. Hope it will get fixed soon!

Thank for your support Dear.

**Comment by htinlinnzaw** [comment]
> I will just copy-paste my answer here, as users may find the answer faster like this:
> 
> I don't know, if the developer of flutter_web_auth is still actively developing his project. For the time being, you can just use my patch using a override, like this (add it to your pubspec.yaml):
> 
> ```
> dependency_overrides:
>   flutter_web_auth:
>     git: 'https://github.com/ThexXTURBOXx/flutter_web_auth.git'
> ```
> 
> To fix this issue using my patch, just pass the additional `saveHistory` parameter with the value `false` to the `authenticate` function:
> 
> ```
> FlutterWebAuth.authenticate(url: url, callbackUrlScheme: scheme, saveHistory: false);
> ```
> 
> Please be aware, that compile issues complaining about `duplicate files` may arise. I don't know exactly, why Dart/Flutter is too dumb to see, that my library should override all files of the other one (if someone has a fix for this problem, you are invited to tell me and all the others :) ).
> My workaround for this is to just delete these folders:
> `$FLUTTER_DIR$\.pub-cache\git\flutter_web_auth-<SOMEHASH>`
> `$FLUTTER_DIR$\.pub-cache\hosted\pub.dartlang.org\flutter_web_auth-<SOMEHASH>`
> and then `pub get` again.

Thank you Dear. It **works** for me.

**Comment by okrad** [comment]
Thank you @ThexXTURBOXx, and thank you @htinlinnzaw for trying out the patch! Did you have the chance to try it out with iOS too?

I forked the flutter_web_auth plugin and applied the patch. 
I'm going to test it out and if everything goes well I'll probably override the import (for the time being, at least until the patch gets integrated in the main repo).

**Comment by ThexXTURBOXx** [comment]
No problem!
My patch only adds the additional, optional parameter, which should not break anything on iOS's side. If the parameter isn't specified, Android just ignores it. I will take a look at the iOS-sided code again and will make sure, I don't break anything.
I can't confirm it 100% though, because I don't have an iOS device and I am not part of the developer program.

**Comment by ThexXTURBOXx** [comment]
Okay, so I don't see, why it should break anything on iOS. I don't know Swift so well, but as I can see, there is a ``Plugin``, which ``handle``s the call to the method ``authenticate`` (much like on Android).
The argument ``FlutterMethodCall call`` then contains the arguments given by ``flutter_web_auth`` as a ``Dictionary<String, AnyObject>``.
Since my patch only affects Android, the Dictionary only cares about the entries ``url`` and ``callbackUrlScheme`` and not my own optional parameter ``saveHistory``.
Therefore, my patch should neither affect any apps that don't use my parameter nor any iOS devices even when the parameter is passed.

**Comment by okrad** [comment]
I agree with you, it shouldn't break anything on iOS... 
I made some more tests without any issues, so I'm going to publish the new version with the overridden plugin...
Thank you again!

**Comment by okrad** [comment]
Problem is, I can't release the new version on pub.dev, since git dependencies are not allowed in published packages...
I think I'll wait a few more days to see if your pull request gets merged in the main repo... 
In the meantime I'll try to contact the author and see if he is still willing to maintain the plugin.

**Comment by ThexXTURBOXx** [comment]
Oh, ah yeah, that makes sense... Keep us up-to-date :)

**Comment by htinlinnzaw** [comment]
I also hoping the new version fixing that issue. Very nice guys and thank for you support.

**Comment by ibelz** [comment]
Any news to this topic?

Ok, the fix with the dependency override from ThexXTURBOXx works for me, but it's really "hacky"... 
The PR at FlutterWebAuth is open since months without any merge. 

Is there any other "offical" solution in sight here?

**Comment by ThexXTURBOXx** [comment]
@ibelz Actually, it's not a hacky solution as it is an old version of the official one. I would really encourage everyone to use the official solution now as it also fixes the same bug for Apple users:
```yaml
dependency_overrides:
  flutter_web_auth:
    git:
      url: 'https://github.com/LinusU/flutter_web_auth.git'
      ref: 'ephemeral'
```

**Comment by ibelz** [comment]
@ThexXTURBOXx 
Oh wow! That sounds great! Much better. 
Thank you very much for your work an help :)

**Comment by Sese-Schneider** [comment]
I've created a convenience fork of this project which can be used to achieve the behavior.

```
dependency_overrides:
  oauth2_client:
    git:
      url: https://github.com/simpleclub-extended/oauth2_client
      ref: ephemeral
```

Hope this helps you!

I will not create a PR for now, because I think @teranetsrl should not merge this until https://github.com/LinusU/flutter_web_auth/pull/64 is available stable on pub.dev

/cc @ibelz @ThexXTURBOXx @DanielEberl

**Comment by AndrewStickler** [comment]
flutter_web_auth now supports ephemeral sessions as of v0.3.1 - any chance of oauth2_client making use of this soon please?

**Comment by ThexXTURBOXx** [comment]
@AndrewStickler I opened a PR (see #96) that supports this feature. Let's hope it gets merged soon.
